File: load-test-framework/python_src/clients/flow_control/rate_limiter_flow_controller.py
from clients.flow_control import FlowController
from concurrent.futures import Executor, ThreadPoolExecutor
import time
from threading import Condition


class RateLimiterFlowController(FlowController):
    """
    A FlowController that allows actions at a given per second rate.
    """
    def __init__(self, per_second_rate: float):
        self.seconds_per_token = 1/per_second_rate
        self.tokens = 0
        self.condition = Condition()
        self.executor: Executor = ThreadPoolExecutor(max_workers=1)
        self.executor.submit(self._token_generator)

    def _token_generator(self):
        time.sleep(self.seconds_per_token)
        with self.condition:
            self.tokens += 1
            self.condition.notify()
        self._token_generator()

    def request_start(self):
        with self.condition:
            while self.tokens < 1:
                self.condition.wait()
            self.tokens -= 1

    def inform_finished(self, was_successful: bool):
        pass

File: load-test-framework/src/main/java/com/google/pubsub/clients/flow_control/RateLimiterFlowController.java
package com.google.pubsub.clients.flow_control;

import com.google.common.util.concurrent.RateLimiter;

/**
 * A FlowController that delegates to a guava RateLimiter for a static allowed rate.
 */
public class RateLimiterFlowController implements FlowController {
    private final RateLimiter limiter;

    public RateLimiterFlowController(double rate) {
        this.limiter = RateLimiter.create(rate);
    }

    @Override
    public int requestStart() {
        limiter.acquire();
        return 1;
    }

    @Override
    public void informFinished(boolean wasSuccessful) {
    }
}
